internal/clock/simulated_clock_test.go

```diff
@@ -0,0 +1,279 @@
+		t.Run(tc.name, func(t *testing.T) {
+			clock := NewSimulatedClock(referenceTime)
+			tc.initialTimeSetup(clock)
+
+			clock.SetTime(tc.timeToSet)
+
+			assert.True(t, clock.Now().Equal(tc.expectedAfter), "After SetTime, clock.Now() was %v, expected %v", clock.Now(), tc.expectedAfter)
+		})
+	}
+}
+
+func TestSimulatedClock_AdvanceTime(t *testing.T) {
+	testCases := []struct {
+		name         string
+		initialTime  time.Time
+		advanceBy    time.Duration
+		expectedTime time.Time
+	}{
+		{
+			name:         "AdvancePositiveDuration",
+			initialTime:  referenceTime,
+			advanceBy:    5 * time.Minute,
+			expectedTime: referenceTime.Add(5 * time.Minute),
+		},
+		{
+			name:         "AdvanceNegativeDuration",
+			initialTime:  referenceTime,
+			advanceBy:    -2 * time.Hour,
+			expectedTime: referenceTime.Add(-2 * time.Hour),
+		},
+		{
+			name:         "AdvanceByZeroDuration",
+			initialTime:  referenceTime,
+			advanceBy:    0,
+			expectedTime: referenceTime,
+		},
+		{
+			name:         "AdvanceFromZeroTime",
+			initialTime:  time.Time{},
+			advanceBy:    time.Hour,
+			expectedTime: (time.Time{}).Add(time.Hour),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			clock := NewSimulatedClock(referenceTime)
+			clock.SetTime(tc.initialTime) // Set initial time for the clock
+
+			clock.AdvanceTime(tc.advanceBy)
+
+			assert.True(t, clock.Now().Equal(tc.expectedTime), "After AdvanceTime, clock.Now() was %v, expected %v", clock.Now(), tc.expectedTime)
+		})
+	}
+}
+
+func TestSimulatedClock_After_ShouldFire(t *testing.T) {
+	testCases := []struct {
+		name          string
+		afterDuration time.Duration
+		action        func(sc *SimulatedClock) // Action to manipulate the clock after After() is called
+	}{
+		{
+			name:          "ZeroDuration_FiresImmediately",
+			afterDuration: 0,
+			action:        func(sc *SimulatedClock) { /* No action needed for immediate fire */ },
+		},
+		{
+			name:          "NegativeDuration_FiresImmediately",
+			afterDuration: -5 * time.Second,
+			action:        func(sc *SimulatedClock) { /* No action needed for immediate fire */ },
+		},
+		{
+			name:          "PositiveDuration_Fires_WhenTimeAdvancedPastDuration",
+			afterDuration: 10 * time.Second,
+			action: func(sc *SimulatedClock) {
+				sc.AdvanceTime(15 * time.Second) // Advance well past the duration
+			},
+		},
+		{
+			name:          "PositiveDuration_Fires_WhenTimeSetPastDuration",
+			afterDuration: 10 * time.Second,
+			action: func(sc *SimulatedClock) {
+				sc.SetTime(referenceTime.Add(15 * time.Second)) // Set time well past the duration
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			clock := NewSimulatedClock(referenceTime)
+			clockTimeAtAfterCall := clock.Now()
+
+			ch := clock.After(tc.afterDuration)
+			require.NotNil(t, ch, "Channel should not be nil")
+
+			// Perform the action (if any) that might trigger the timer
+			if tc.action != nil {
```

Comment:
Let's remove this check if possible - I don't see this function being set to nil and the lesser the number of conditionals in test code, the better it is.